### ISSUE TITLE
Cypress aliases are not refreshed before assertion

### ISSUE THREAD

**Issue by TomaszG** [cypress-io/cypress#5153]
### Current behavior:

Cypress aliases refer to old DOM state when they're fetched/asserted in the test which leads to the assertions like below passing:
```javascript
(...)
cy.get('@rows')
  .should('have.length', 2)

  .get('@rows')
  .should('have.length', 3);
```

Apparently first assertion is passing because of the cached DOM, and second one passes because alias gets refreshed as first try fails silently

### Desired behavior:

Aliases should refer to the actual DOM state when they're used in the test.

### Steps to reproduce:

1. clone https://github.com/TomaszG/cypress-issues
2. `npm i`
2. `npm run start`
3. `npx cypress run`

### Versions

Cypress 3.4.1
Chromium 78
macOS 10.14.6

**Comment by ryan-snyder** [comment]
Fix for this would be to do
```js
/// <reference types="Cypress" />

describe('Aliases test', () => {
  beforeEach(() => {
    cy.visit('/')
  })
  it('should not pass', () => {
    cy.get('tbody').as('rows')
    // when
    cy.get('@rows').children().should('have.length', 2)
    cy.get('#addbtn').click();
    cy.get('@rows').children().should('have.length', 3);
  })
})
```
instead.

So you are correct in that cypress does not query the dom again but from the docs that seems to be expected.
From https://docs.cypress.io/guides/core-concepts/variables-and-aliases.html#Elements
`Because we’ve used the @ character in cy.get(), instead of querying the DOM for elements, cy.get() looks for an existing alias called rows and returns the reference (if it finds it).`
The only case where cypress will query the dom again is ` If you alias DOM elements that have been removed from the DOM by the time you call cy.get() with the alias, Cypress automatically re-queries the DOM to find these elements again.`
In other words, because you are appending a new element to your existing table and the DOM elements the alias is reference are still in the DOM, cypress returns the alias reference that only had two rows.

Hope that makes sense!

**Comment by TomaszG** [comment]
Well, I wouldn't call it a fix, rather workaround which may not be valid for all scenarios as, e.g. in my tests I prefer pointing aliases at the elements which number is changing dynamically (and can have various parents). Also, if the alias is a reference pointing at some DOM elements then I believe it should reflect the current DOM state. What's more, there's some inconsistency in the behaviour - if I had removed the row instead of adding it, the alias would be working fine, is that right?

If this cannot be easily fixed maybe it's worth to provide users an option to refresh the alias as an option for the `get('@alias')` method?

**Comment by ryan-snyder** [comment]
Perhaps someone from the cypress dev team can answer this but you are correct that the behaviour is inconsistent. I would expect it to work the same whether you're adding or removing. I still believe the reason is that your alias is referencing the individual `tr` elements. I want to say the reason that cypress doesn't re-query the dom when adding a tr is because cypress has no way to tell that the tr that was added should be a part of the alias. Because like the documentation says, because you've used the `@` character, it gets the existing alias.
@bahmutov Am I understanding this right?

Regardless, I would say that this is more of a feature request than a bug but I could be wrong

**Comment by ryan-snyder** [comment]
@jennifer-shehane Proposal to close this ticket and make a feature request to force a DOM re-query on a get? Perhaps something like `cy.get('@alias', { refresh: true })`?

**Comment by TomaszG** [comment]
It seems to be quite similar to that one https://github.com/cypress-io/cypress/issues/2971. Do you know how I can get a selector from alias? I was thinking about overwriting get() command and add to it an option to fetch the DOM using selector referenced by alias.

**Comment by ryan-snyder** [comment]
It is similar to that one. You should be able to get the selector by doing 
```js
cy.get().then(ele => {
ele.selector
}
```
Or something similar. If you do a console.log on ele, it should be pretty easy to do

**Comment by TomaszG** [comment]
So I have the following piece of code which fails because of the issue mentioned in the ticket (previously @errorMessages alias points to 2 elements, however, at this point there's only one such element)
```js
(...)
  // then
  .get('@errorMessages')
  .should('have.length', 1)
  .and('have.text', missingLabelErrorMessage)
```

According to your advice, I tried to try the workaround in the test which worked fine:
```js
  // then
  .get('@errorMessages').then(element => {
    cy.get(element.prevObject.selector)
      .should('have.length', 1)
      .and('have.text', missingLabelErrorMessage)
    })
```

So I tried to make it generic:
```js
Cypress.Commands.overwrite('get', (originalFn, selector, options) => {
  if (selector.startsWith('@') && options && options.refereshAlias) {
    cy.get(selector).then(element => originalFn(element.prevObject.selector, options));
  }
  return originalFn(selector, options);
});
```

However, now when I run the test, even without providing `refereshAlias` options (so code inside the `if` instruction is not executed) I get the following error during execution one of `get('@someOtherAlias')` command:

> CONTAINS Object{5}
> CypressError: cy.contains() can only accept a string, number or regular expression.

which doesn't make sense, as error comes up from `get('@someOtherAlias')` command (checked with `debug()`)

**Comment by jennifer-shehane** [comment]
We currently don't requery the DOM for aliased elements, as noted above from https://docs.cypress.io/guides/core-concepts/variables-and-aliases.html#Elements

I would consider this a duplicate of the proposal in https://github.com/cypress-io/cypress/issues/2971 Can we close this issue in favor of that please?

**Comment by TomaszG** [comment]
Sure thing, closing it now, thanks!
